File: lab7/Python/philosopher_greedy.py
import threading
from time import sleep, time_ns


class PhilosopherGreedy(threading.Thread):
    def __init__(self, num, locks, forks):
        threading.Thread.__init__(self)
        self.num = num
        self.locks = locks
        self.forks = forks
        self.left_fork = None
        self.right_fork = None
        self.time = 0

    @staticmethod
    def think():
        sleep(0.001)

    @staticmethod
    def eat():
        sleep(0.001)

    def run(self):
        i = 0
        while i < 100:
            self.think()

            left = self.num
            right = self.num + 1
            if right >= len(self.locks):
                right = 0

            start_waiting = time_ns()
            if self.locks[left].acquire(False):
                if self.locks[right].acquire(False):
                    stop_waiting = time_ns()
                    self.time += (stop_waiting - start_waiting)

                    self.left_fork = self.forks[left]
                    self.right_fork = self.forks[right]

                    self.eat()
                    i += 1

                    self.locks[right].release()
                    self.locks[left].release()

File: lab7/Java/src/com/company/PhilosopherGreedy.java
package com.company;

import java.util.concurrent.locks.ReentrantLock;

public class PhilosopherGreedy implements Runnable
{
    private int num;  // current philosopher number

    // for i-th philosopher:
    // left fork = i-th fork
    // right fork = (i+1)-th fork
    private ReentrantLock[] locks;
    private Object[] forks;

    private Object leftFork = null;
    private Object rightFork = null;

    // waiting time - return value
    long time = 0;

    public PhilosopherGreedy(int num, ReentrantLock[] locks, Object[] forks)
    {
        this.num = num;
        this.locks = locks;
        this.forks = forks;
    }

    private void think() throws InterruptedException
    {
        Thread.sleep(1);
    }


    private void eat() throws InterruptedException
    {
        Thread.sleep(1);
    }

    @Override
    public void run()
    {
        try
        {
            int i = 0;
            while (i < 100)
            {
                think();

                int left = num;
                int right = num + 1;
                if (num == locks.length - 1)
                    right = 0;

                long startWaiting = System.nanoTime();
                if (locks[left].tryLock())
                {
                    try
                    {
                        if (locks[right].tryLock())
                        {
                            try
                            {
                                long stopWaiting = System.nanoTime();
                                this.time += (stopWaiting - startWaiting);

                                this.leftFork = forks[left];
                                this.rightFork = forks[right];

                                eat();
                                i += 1;
                            }
                            finally
                            {
                                this.rightFork = null;
                                locks[right].unlock();
                            }
                        }
                    }
                    finally
                    {
                        this.leftFork = null;
                        locks[left].unlock();
                    }
                }
            }
        }
        catch (InterruptedException ignored)
        { }
    }
}